### PR TITLE
allow to manually flush the mount cache (fixes #803)

### LINKED ISSUE #803: Manual trigger of cache flush
The updates to fix issue #680 are a huge help to refresh the cache without having to unmount/remount, so thanks for that!

However, it would be very useful to also be able to manually trigger a refresh of the cache in instances where we already know the cache is stale.

Issue #771 discusses having an rclone process you can send commands to; perhaps a refresh/flush command could be implemented if/when that functionality is added?

### LINKED ISSUE #1375: mount looks up wrong path when parent directories were renamed
> What is your rclone version (eg output from `rclone -V`)

bc25190fc719f9c858d9ed06eb85444a45f20d80 (+some patches, but not relevant)

> Which cloud storage system are you using? (eg Google Drive)

GDrive, but I believe this is not specifically related to it.

steps to reproduce (although I did not check if they can be shorter):
```
rclone mount -vvv somefs:/ ~/test
cd ~/test


stefan@aard ~/test $ mkdir -p a/b/c
stefan@aard ~/test $ touch a/b/c/file
stefan@aard ~/test $ ls -1 a  
b
stefan@aard ~/test $ ls -1 a/b
c
stefan@aard ~/test $ ls -1 a/b/c
file
stefan@aard ~/test $ mv a/b a/bb
stefan@aard ~/test $ ls -1 a/
bb
stefan@aard ~/test $ ls -1 a/bb/
c
stefan@aard ~/test $ ls -1 a/bb/c/
# no file shows up here!
```

The log excerpt is most interesting at the end:

```
2017/04/29 18:00:53 DEBUG : a: Dir.Mkdir
2017/04/29 18:00:54 DEBUG : a: Dir.Mkdir OK
2017/04/29 18:00:54 DEBUG : a: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:00:54 DEBUG : a/b: Dir.Lookup
2017/04/29 18:00:54 DEBUG : a: Reading directory
2017/04/29 18:00:54 DEBUG : Google drive root '': Reading "a/"
2017/04/29 18:00:54 DEBUG : Google drive root '': Finished reading "a/"
2017/04/29 18:00:54 DEBUG : a/b: Dir.Mkdir
2017/04/29 18:00:55 DEBUG : a/b: Dir.Mkdir OK
2017/04/29 18:00:55 DEBUG : a/b: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:00:55 DEBUG : a/b/c: Dir.Lookup
2017/04/29 18:00:55 DEBUG : a/b: Reading directory
2017/04/29 18:00:55 DEBUG : Google drive root '': Reading "a/b/"
2017/04/29 18:00:55 DEBUG : Google drive root '': Finished reading "a/b/"
2017/04/29 18:00:55 DEBUG : a/b/c: Dir.Mkdir
2017/04/29 18:00:57 DEBUG : a/b/c: Dir.Mkdir OK
2017/04/29 18:00:57 DEBUG : a/b/c: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:00:57 DEBUG : : Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:00 DEBUG : a/b/c/file: Dir.Lookup
2017/04/29 18:01:00 DEBUG : a/b/c: Reading directory
2017/04/29 18:01:00 DEBUG : Google drive root '': Reading "a/b/c/"
2017/04/29 18:01:01 DEBUG : Google drive root '': Finished reading "a/b/c/"
2017/04/29 18:01:01 DEBUG : a/b/c/file: Dir.Create
2017/04/29 18:01:01 DEBUG : a/b/c/file: Dir.Create OK
2017/04/29 18:01:01 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
2017/04/29 18:01:01 DEBUG : a/b/c/file: WriteFileHandle.Flush
2017/04/29 18:01:01 DEBUG : a/b/c/file: WriteFileHandle.Flush ignoring flush on unwritten handle
2017/04/29 18:01:01 DEBUG : File.Attr valid=1m0s ino=0 size=0 mode=-rw-r--r--
2017/04/29 18:01:01 DEBUG : a/b/c/file: WriteFileHandle.Flush
2017/04/29 18:01:01 DEBUG : a/b/c/file: WriteFileHandle.Flush ignoring flush on unwritten handle
2017/04/29 18:01:01 DEBUG : a/b/c/file: WriteFileHandle.Release closing
2017/04/29 18:01:03 DEBUG : a/b/c/file: File.applyPendingModTime OK
2017/04/29 18:01:03 DEBUG : a/b/c/file: WriteFileHandle.Release OK
2017/04/29 18:01:06 DEBUG : a: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:06 DEBUG : a: Dir.ReadDirAll
2017/04/29 18:01:06 DEBUG : a: Dir.ReadDirAll OK with 1 entries
2017/04/29 18:01:06 DEBUG : a/b: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:08 DEBUG : b: Dir.Lookup
2017/04/29 18:01:08 DEBUG : b: Dir.Lookup
2017/04/29 18:01:12 DEBUG : a/b: Dir.ReadDirAll
2017/04/29 18:01:12 DEBUG : a/b: Dir.ReadDirAll OK with 1 entries
2017/04/29 18:01:12 DEBUG : a/b/c: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:14 DEBUG : a/b/c: Dir.ReadDirAll
2017/04/29 18:01:14 DEBUG : a/b/c: Dir.ReadDirAll OK with 1 entries
2017/04/29 18:01:26 DEBUG : a/bb: Dir.Lookup
2017/04/29 18:01:26 DEBUG : a/b: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:26 DEBUG : a/bb: Dir.Lookup
2017/04/29 18:01:26 DEBUG : a/bb: Dir.Lookup
2017/04/29 18:01:26 DEBUG : a/b: Dir.Rename to "a/bb"
2017/04/29 18:01:27 DEBUG : a/b: Updating dir with a/bb 0xc420298f50
2017/04/29 18:01:27 DEBUG : a/b: Forgetting directory
2017/04/29 18:01:27 DEBUG : a/b/c: Forgetting directory
2017/04/29 18:01:27 DEBUG : a/bb: Dir.Rename renamed from "a/b"
2017/04/29 18:01:30 DEBUG : a: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:30 DEBUG : a: Dir.ReadDirAll
2017/04/29 18:01:30 DEBUG : a: Dir.ReadDirAll OK with 1 entries
2017/04/29 18:01:30 DEBUG : a/bb: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:32 DEBUG : a/bb: Dir.ReadDirAll
2017/04/29 18:01:32 DEBUG : a/bb: Reading directory
2017/04/29 18:01:33 DEBUG : Google drive root '': Reading "a/bb/"
2017/04/29 18:01:33 DEBUG : Google drive root '': Finished reading "a/bb/"
2017/04/29 18:01:33 DEBUG : a/bb: Dir.ReadDirAll OK with 1 entries
2017/04/29 18:01:33 DEBUG : a/b/c: Dir.Attr valid=1m0s ino=0 size=0 mode=drwxr-xr-x
2017/04/29 18:01:36 DEBUG : a/b/c: Dir.ReadDirAll
2017/04/29 18:01:36 DEBUG : a/b/c: Reading directory
2017/04/29 18:01:36 DEBUG : a/b/c: Dir.ReadDirAll OK with 0 entries
```

I suspect that a subdir's inode does not change when a parent directory is renamed. As such, basil/fuse keeps the reference to rclone's `Dir` around, which of course contains the full `path` – this is why it reads `a/b/c` at the end, instead of the expected `a/bb/c`. It is correct on the remote end.

I will probably solve this alongside #803, but here is in case I don't manage.

### PR DESCRIPTION
**Pull request by breunigs**
some comments on this PR:
* I will rebase/squash when this is ready. I expect your review will yield some more changes :)
* fixes #803, and allows fine grained eviction if needed
* I did not need to modify bazil/fuse after all: It already offers a callback through `NodeForgetter` when it discards something on its side. Together with the already implemented rename/remove, this should cover all cases.
* the extra code in `func (d *Dir) rename(…)` is to fix #1375. I experience it when I test manually with GDrive mounted, but I didn't manage to write a failing test case. I'll have to see in #1375 if this is just a bug in GDrive, just not present in the local backend used in test cases or maybe a timing related issue…
* I wonder if SIGHUP is a good signal choice. I would prefer something like SIGUSR1, but if I read [correctly this is not available on Windows](https://github.com/golang/sys/blob/ff24cb3cd86e67842e0d078a5e0cb92d9ffcf8b4/windows/ztypes_windows.go#L51-L84). What about SIGALRM? If not, rclone should probably ignore SIGHUP in other modes, because users might otherwise accidentally kill a running sync if they pass the wrong PID.

### CODE REVIEW

File: cmd/mount/dir.go [hunk 1]
```diff
@@ -16,8 +18,8 @@ import (
 )
 
 var (
-	dirLastReadTimes      = make(map[string]time.Time)
-	dirLastReadTimesMutex sync.RWMutex
+	cachedDirs      = make(map[string]*Dir)
```
Comment (ncw) [inline cmd/mount/dir.go]: I think the cachedDirs map is probably unnecessary.  The Dir tree has all that info in already and with a ForgetAll method you can recurse calling the ForgetAll method of child directories very easily.

**Comment by ncw** [comment]
Re SIGHUP:
From [the signal docs](https://golang.org/pkg/os/signal/#hdr-Windows) I think only Interrupt is supported under Windows.  Anyway the whole of fuse is compiled out for windows at the moment.

SIGHUP usually means reload the config file so I wonder whether SIGUSR1 might be a better choice.  SIGUSR2 means kick the bandwidth limiter at the moment too.

**Comment by breunigs** [comment]
Ah, I will take a look at the dir tree you mentioned. That sounds like exactly like what I need, except it comes without the double maintenance. 

Regarding reviewing non-squashed patches: I am curious, how do you review usually? Github has the "Files Changed" tab, which essentially works like a squashed view. Locally, something like `git diff master...` should do the trick (the three dots compare it to when it was branched off, not the current state of the `master` branch). Regardless, I will squash in the future.

**Comment by breunigs** [comment]
Your approach works much better! The updated patch still needs a lot of cleanup though.

However, do you remember why `d.items = nil` was needed when renaming a directory? It is there since this feature was supported, but the test I wrote for this corner case passes either way. I also tried with a shell, which keeps the reference to the current directory without a traversal. They also behave the same, as far as I can tell.

File: cmd/mount/fs.go [hunk 2]
```diff
@@ -15,20 +18,26 @@ import (
 
 // FS represents the top level filing system
 type FS struct {
-	f fs.Fs
+	f       fs.Fs
+	rootDir *Dir
 }
 
+var myFirstFs *FS
```
Comment (ncw) [inline cmd/mount/fs.go]: This is a little inelegant, but I guess it is OK for testing!

**Comment by ncw** [comment]
> Regarding reviewing non-squashed patches: I am curious, how do you review usually? Github has the "Files Changed" tab, which essentially works like a squashed view. Locally, something like git diff master... should do the trick (the three dots compare it to when it was branched off, not the current state of the master branch). Regardless, I will squash in the future.

I don't think I ever clicked on that tab before - thanks for teaching me a new trick :-)  From previous revision control systems all the way back to patches to a mailing list I've always reviewed individual diffs and I guess I just carried that forward to github!

> However, do you remember why d.items = nil was needed when renaming a directory? It is there since this feature was supported, but the test I wrote for this corner case passes either way. I also tried with a shell, which keeps the reference to the current directory without a traversal. They also behave the same, as far as I can tell.

It was there to get rid of any items in a directory.  Each rclone Object contains within it the full path in the remote.  This is great for syncing etc, but for a file system when you rename the directory you need to invalidate all those objects because they now have the wrong paths in them.  Directory listings etc will work without `d.items = nil` but if you try to read any of the data it will likely say object not found (on bucket based remotes it will certainly (eg swift/s3/b2), on drive, acd I guess it may not!).

I'll just note that I've been refactoring the underlying directory tree / nodes into [mountlib](https://github.com/ncw/rclone/tree/cmount/cmd/mountlib) (currently in the [cmount](https://github.com/ncw/rclone/tree/cmount) branch) so I can re-use it with a different fuse library ([cgofuse](https://github.com/billziss-gh/cgofuse)) which can be made to work under windows.  I'd also like to use the mountlib to serve webdav also.

Anyway at some point this means I'll need to merge what you've done into the mountlib (that isn't a problem - I've done that sort of thing lots of times before!).

I'd appreciate your feedback on mountlib if you have a moment!  At some point I'd like to make some tests for mountlib itself rather than going through FUSE but I haven't done that yet!

**Comment by ncw** [comment]
BTW there is a golint message to fix!

```
/home/travis/gopath/src/github.com/ncw/rclone/cmd/mount/dir.go:75:1: comment on exported method Dir.ForgetAll should be of the form "ForgetAll ..."
```

**Comment by breunigs** [comment]
1. I've added "checkout mountlib" to my todo
2. thanks for the hint about `items = nil`. I will see if I can make it fail locally, and maybe even write a test for it.
3. regarding changes: I wanted to check how gdrive/ACD changes APIs actually work. Right now the more specific `ForgetPath` is only used in tests, so I might at least make it suitable for implementing the "listen to changes" for these remotes. 
I still need to at least roughly understand how the mount and the additional "changes listener" interface for remotes would work together. The mount polling via `mount.FS.f.somehowGetChanges` would probably work, but maybe there is a more elegant solution.
Last but not least, I wanted to look into getting rid of the `myFirstFs` hack.

**Comment by breunigs** [comment]
I guess this is as good as it gets, if you wanna take another look.

Both Google and Amazon's changes API allow one to update the cache – so ideally those APIs shouldn't need to call "Forget", but rather implement some custom update function that updates the file in the existing caches. The generic `walk` function, like you originally suggested should help with that.

**Comment by ncw** [comment]
That looks great thank you :-)

**Comment by ncw** [comment]
I forward ported this to the cmount branch - can you give it a quick once over and see if it looks sane? add408c9edb7954d2d749981fb516de31bdcbe53

Thanks

**Comment by breunigs** [comment]
I gave the branch a test spin and SIGHUP'ing rclone still works nicely. I missed `dirCacheTime` being hardcoded when cross reading the changeset though, and I was quite stumped for a while why `--dir-cache-time 10s` didn't work. On the upside, I am now quite sure the cache time invalidation also still works.

I took a look at the diff and I must say you have been quite busy. It is not yet quite clear to me what's the idea for the code organization. Is `mountlib` supposed to be a generic base that `mount` (bazil/fuse) and `cmount` (cgofuse) can share? And both bazil and cgofuse will remain in rclone for the forseeable future?

These are the things popping into my mind. All rather minor though:
- Some of the pieces of code in `cmount` and `mount` seem very similar. I wonder if there's a possibility to extract more shared code into `mountlib` or similar. I would have to have a much deeper look though to see if this would actually be better. For instance, consider the `waitloop:`. Sharing that would work, as long as one doesn't call `signal.Notify(sigInt, syscall.SIGINT, syscall.SIGTERM)` in the cgofuse path. Hard to tell for me if the if/else would be better than the code duplication. Options that are supported by both would be another example.
- I'll have to actually try to contribute a patch into the `cmount` branch to see if I can get comfortable with the file naming and package naming: now there's `fs.Fs`, `mount.FS`, `cmount.FS`, `mountlib.FS` and `mounttest.FS` and they even call into each other. It was already a source of confusion for me when there were only two.
- if a more stable inode allocation is needed, bazil/fuse has actually a pretty nice implementation: https://github.com/bazil/fuse/blob/master/fs/serve.go#L1551 Implementing bazile/fuse's `FSInodeGenerator` using rclone's `NewInode` would probably also solve your fixmes in `cmd/mount/dir.go ReadDirAll`. 

That's it for now. If something else comes to mind, I'll add it here.
